src/krb5_plugin/passkey/passkey_utils.c

```diff
@@ -0,0 +1,556 @@
+/*
+    Authors:
+        Pavel Březina <pbrezina@redhat.com>
+
+    Copyright (C) 2021 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "config.h"
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <jansson.h>
+#include <arpa/inet.h>
+#include <krb5/preauth_plugin.h>
+
+#include "krb5_plugin/common/utils.h"
+#include "krb5_plugin/passkey/passkey.h"
+
+void
+sss_passkey_config_free(struct sss_passkey_config *passkey)
+{
+    if (passkey == NULL) {
+        return;
+    }
+
+    sss_string_array_free(passkey->indicators);
+    free(passkey);
+}
+
+/**
+ * {
+ *   "indicators": ["..."] (optional)
+ * }
+ */
+krb5_error_code
+sss_passkey_config_init(const char *config,
+                        struct sss_passkey_config **_passkey)
+{
+    struct sss_passkey_config *passkey;
+    json_t *jindicators = NULL;
+    json_error_t jret;
+    json_t *jroot;
+    krb5_error_code ret;
+
+    passkey = malloc(sizeof(struct sss_passkey_config));
+    if (passkey == NULL) {
+        return ENOMEM;
+    }
+    memset(passkey, 0, sizeof(struct sss_passkey_config));
+
+    jroot = json_loads(config, 0, &jret);
+    if (jroot == NULL) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = json_unpack(jroot, "{s?:o}", "indicators", &jindicators);
```

Comment:
FreeIPA puts all indicator strings in `[...]`, see https://pagure.io/freeipa/blob/master/f/daemons/ipa-kdb/ipa_kdb_principals.c#_619, so `"[{s?:o}]"` should be used here as well.

Reply:
otp is required to be array by Kerberos since there is support for multiple token.

For idp, we wanted to use array as well since there may be multiple idp mechanisms in the future.

I think for passkey, it is sufficient to be a single value since it implements a specific protocol.

Reply:
Hi,

if you prefer it this way you have to coordinate with @flo-renaud and @abbra since currently the array syntax is used at https://github.com/flo-renaud/freeipa/blob/passkey/daemons/ipa-kdb/ipa_kdb_principals.c#L647.

bye,
Sumit

Reply:
FYI @flo-renaud @abbra what is your opinion on this?

Reply:
I'm with @sumit-bose. Let us keep the syntax the same for all those strings on IPA side:
```
    const krb5_octet rad_string[] = "otp\0[{\"indicators\": [\"radius\"]}]";
    const krb5_octet otp_string[] = "otp\0[{\"indicators\": [\"otp\"]}]";
    const krb5_octet idp_string[] = "idp\0[{\"type\":\"oauth2\",\"indicators\": [\"idp\"]}]";
    const krb5_octet passkey_string[] = "passkey\0[{\"indicators\": [\"passkey\"]}]";
```

Reply:
A bit of reasoning: I have plans on extending the speicific indicators in future to include details that might be useful for the applications which would authorize access based on the authentication indicators. This is to not exclude level of assurance (RFC 6711) and similar methods specifically mentioned in the authentication indicator's RFC 8129.

Reply:
Ok, I changed it.